### PR TITLE
make autoproj_bootstrap and autoproj_install handle --no-interactive

### PR DESCRIPTION
**Pull request by doudou**

### CODE REVIEW

File: lib/autoproj/ops/main_config_switcher.rb [hunk 1]
```diff
@@ -16,16 +16,21 @@ def initialize(ws)
             EXPECTED_ROOT_ENTRIES = [".", "..", "autoproj_bootstrap",
                                      ".autoproj", "bootstrap.sh", ENV_FILENAME].to_set
 
+
+            def interactive_flag_from_env
```
Comment (2maz) [inline lib/autoproj/ops/main_config_switcher.rb]: would expect this to be "ignore_empty_dir_from_env"  while interactive_flag_from_env should refer to AUTOPROJ_NONINTERACTIVE

File: lib/autoproj/ops/main_config_switcher.rb [hunk 2]
```diff
@@ -16,16 +16,21 @@ def initialize(ws)
...
+
             # Verifies that {#root_dir} contains only expected entries, to make
             # sure that the user bootstraps into a new directory
             #
             # If the environment variable AUTOPROJ_BOOTSTRAP_IGNORE_NONEMPTY_DIR
             # is set to 1, the check is skipped
-            def check_root_dir_empty
-                return if ENV['AUTOPROJ_BOOTSTRAP_IGNORE_NONEMPTY_DIR'] == '1'
+            def check_root_dir_empty(interactive: interactive_flag_from_env)
+                return unless interactive
```
Comment (2maz) [inline lib/autoproj/ops/main_config_switcher.rb]: linked to comment above: why not check 'ignore_nonempty_dir' or 'noninteractive' ?

File: lib/autoproj/cli/main.rb [hunk 3]
```diff
@@ -120,15 +133,15 @@ def bootstrap(*args)
...
                     ops.run
-                    exec Gem.ruby, $0, 'bootstrap', *args
+                    exec Gem.ruby, $0, 'bootstrap', *bootstrap_options
                 end
                 run_autoproj_cli(:bootstrap, :Bootstrap, Hash[], *args)
             end
 
             desc 'envsh', 'update the environment files'
-            def envsh
+            def envsh(*args)
```
Comment (2maz) [inline lib/autoproj/cli/main.rb]: argument looks used?

File: lib/autoproj/cli/envsh.rb [hunk 4]
```diff
@@ -4,7 +4,7 @@ module CLI
         class Envsh < InspectionTool
             def validate_options(_unused, options = Hash.new)
                 _, options = super(_unused, options)
-                options
+                [options]
```
Comment (2maz) [inline lib/autoproj/cli/envsh.rb]: are there already tests for cli/test_envsh?

File: test/ops/test_install.rb [hunk 5]
```diff
@@ -25,6 +25,23 @@ module Ops
                     gemfile_source: gemfile_source
             end
 
+            it "may install non-interactively" do
```
Comment (2maz) [inline test/ops/test_install.rb]: ruby test/ops/test_install.rb fails 
```
Errno::ENOENT: No such file or directory @ dir_s_mkdir  .../autoproj/lib/autoproj/../../vendor/test_gem_home
    /usr/lib/ruby/2.5.0/fileutils.rb:1265:in `mkdir'
   ...
```
anything i am missing?

File: lib/autoproj/ops/main_config_switcher.rb (hunk 1 above)
Comment (doudou) [reply lib/autoproj/ops/main_config_switcher.rb]: I didn't recall the existence of `AUTOPROJ_NONINTERACTIVE`. I'd deprecate AUTOPROJ_BOOTSTRAP_IGNORE_NONEMPTY_DIR in favor of `NONINTERACTIVE` to be honest. Fine with you if I test for both ?

File: lib/autoproj/ops/main_config_switcher.rb (hunk 2 above)
Comment (doudou) [reply lib/autoproj/ops/main_config_switcher.rb]: I don't understand this comment ... the `interactive_flag_from_env` method is here to abstract exactly that. It's named `interactive` as that's the name of the flag.

File: test/ops/test_install.rb (hunk 5 above)
Comment (doudou) [reply test/ops/test_install.rb]: It seems that the tests assume you've checked out your autoproj installation separately, and bundle-installed it with `--path=vendor`.

Could you try it this way ? If that solves the issue, I'll add some documentation in README for it.

File: lib/autoproj/ops/main_config_switcher.rb (hunk 1 above)
Comment (2maz) [reply lib/autoproj/ops/main_config_switcher.rb]: Sure

File: lib/autoproj/ops/main_config_switcher.rb (hunk 2 above)
Comment (2maz) [reply lib/autoproj/ops/main_config_switcher.rb]: If you deprecate the AUTOPROJ_BOOTSTRAP_IGNORE_NONEMPTY_DIR and also check for AUTOPROJ_NONINTERACTIVE then fine.
I mean that you can also set the env var to ignore non empty dir, but that does not necessarily mean you bootstrap noninteractive.

File: test/ops/test_install.rb (hunk 5 above)
Comment (2maz) [reply test/ops/test_install.rb]: Sorry, not clear how I should properly test now.
Wouldn't ensuring the existence of the directory by fixing Test::fixture_gem_home in lib/autoproj/test.rb be sufficient? 
If doing that the test starts running,  but still gives error: 

```
  1) Error:
Autoproj::Ops::Install::default shared gems location#test_0004_sets the environment to point RubyGems to the shared location:
Gem::UnsatisfiableDependencyError: Unable to resolve dependency: user requested 'did_you_mean (= 1.2.0)'
```

Comment (doudou) [reply test/ops/test_install.rb]: The code assumes that the gem location is in `vendor/`

Which happens if you install autoproj (for testing) with bundler, that is:

~~~
git clone https://github.com/rock-core/autoproj
cd autoproj
bundle install --path=vendor
bundle exec ruby test/ops/test_install.rb
~~~

File: lib/autoproj/cli/envsh.rb (hunk 4 above)
Comment (doudou) [reply lib/autoproj/cli/envsh.rb]: Added one

File: lib/autoproj/ops/main_config_switcher.rb (hunk 1 above)
Comment (doudou) [reply lib/autoproj/ops/main_config_switcher.rb]: In the end I followed your original idea, and kept IGNORE_NONEMPTY. You just have NONINTERACTIVE == '1' that implies IGNORE_NONEMPTY

File: lib/autoproj/cli/main.rb (hunk 3 above)
Comment (doudou) [reply lib/autoproj/cli/main.rb]: fixed

File: lib/autoproj/cli/envsh.rb (hunk 4 above)
Comment (doudou) [reply lib/autoproj/cli/envsh.rb]: added one

**Comment by doudou** [comment]
fixed your comments, and added more system tests regarding interactivity during bootstrap (i.e. actually calling the CLI as a subprocess)

**Comment by 2maz** [comment]
Thx, changes look good to me.

I failed to run the tests however - precise test instructions might help:
```
Run options: --seed 49622

# Running:

ERROR:  Could not find a valid gem 'bundle' (= 2.0.1) in any repository
ERROR:  Possible alternatives: bundle
Detected 'gem' to be /usr/bin/gem2.5
Installing bundler in /tmp/d20190403-31219-gabntj/ruby/2.5.0
+ /usr/bin/ruby2.5 -e load ".../workspace/test_autoproj/autoproj_new/bin/autoproj_install" -- --gems-path=/tmp/d20190403-31219-gabntj --seed-config .../workspace/test_autoproj/autoproj_new/test/scripts/seed-config.yml --gemfile /tmp/d20190403-31219-xkr0ad/Gemfile-dev --gem-source http://localhost:8808
ERROR:  Can't find Ruby library file or shared library bundler/setup
ERROR:  Could not find a valid gem 'bundler' (>= 0) in any repository
FATAL: failed to install bundler in /tmp/d20190403-31219-gabntj/ruby/2.5.0
ERROR:  Can't find Ruby library file or shared library bundler/setup
gem install bundler returned successfully, but still cannot find bundler in /tmp/d20190403-31219-gabntj/ruby/2.5.0
FERROR:  Could not find a valid gem 'bundle' (= 2.0.1) in any repository
ERROR:  Possible alternatives: bundle

```

**Comment by doudou** [comment]
> Thx, changes look good to me.
> 
> I failed to run the tests however - precise test instructions might help:
> 
> ```
> Run options: --seed 49622
> 
> # Running:
> 
> ERROR:  Could not find a valid gem 'bundle' (= 2.0.1) in any repository
> ERROR:  Possible alternatives: bundle
> Detected 'gem' to be /usr/bin/gem2.5
> Installing bundler in /tmp/d20190403-31219-gabntj/ruby/2.5.0
> + /usr/bin/ruby2.5 -e load ".../workspace/test_autoproj/autoproj_new/bin/autoproj_install" -- --gems-path=/tmp/d20190403-31219-gabntj --seed-config .../workspace/test_autoproj/autoproj_new/test/scripts/seed-config.yml --gemfile /tmp/d20190403-31219-xkr0ad/Gemfile-dev --gem-source http://localhost:8808
> ERROR:  Can't find Ruby library file or shared library bundler/setup
> ERROR:  Could not find a valid gem 'bundler' (>= 0) in any repository
> FATAL: failed to install bundler in /tmp/d20190403-31219-gabntj/ruby/2.5.0
> ERROR:  Can't find Ruby library file or shared library bundler/setup
> gem install bundler returned successfully, but still cannot find bundler in /tmp/d20190403-31219-gabntj/ruby/2.5.0
> FERROR:  Could not find a valid gem 'bundle' (= 2.0.1) in any repository
> ERROR:  Possible alternatives: bundle
> ```

Well ... you know ... "it works for me".

It is not really a matter of "better instructions" rather than "having someone else run the tests on his machine".

My goal after the current PRs are merged would be to finally get the test suite pass on Travis before I do a release. Maybe having you get it to run as well would be a worthwhile goal ;-)

**Comment by doudou** [comment]
If you agree with that, I'll merge this (since you seem to be OK with it now). We'll focus on getting the test suite to pass on travis / on your computer afterwards, OK ?

**Comment by 2maz** [comment]
> If you agree with that, I'll merge this (since you seem to be OK with it now). We'll focus on getting the test suite to pass on travis / on your computer afterwards, OK ?

Fine with me.

**Pull request merged**